### DISCUSSION TITLE
Add option to format operators on new line by default

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/roslyn#47601] [General] [answered]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/1167703/add-option-to-format-operators-on-new-line-by-defa.html)._

---
It would be nice if Visual Studio has an option to place operators to the start of a new line instead of putting it at the end. It makes the logical link between variables clear by vertical scanning. It also makes the code look cleaner:

Example:

```
// Please do this - read vertically down :)

var foo = barValue
            && checkBazValue()
            && qux. IsTrue()

this. Character.Location += this. DirectionMode.GetDeltaPoint() 
				* (float)interval. TotalMilliseconds 
				* this. Speed;

this. Key = Enum.TryParse<Key>(key, ignoreCase: true, out var keyboardKey)
    		? keyboardKey
    		: null;

**

// Please stop doing this - scan entire line :(

var foo = barValue &&
            checkBazValue() &&
            qux. IsTrue()

this. Character.Location += this. DirectionMode.GetDeltaPoint() *
				(float)interval. TotalMilliseconds *
				this. Speed;

this. Key = Enum.TryParse<Key>(key, ignoreCase: true, out var keyboardKey) ?
    		keyboardKey :
    		null;
```

---
#### Feedback Bot on 8/31/2020, 03:07 AM: 

Thank you for taking the time to provide your suggestion.&nbsp; We will do some preliminary checks to make sure we can proceed further.&nbsp; We'll provide an update once the issue has been triaged by the product team.

**Comment by sharwell** [comment, marked as answer]
This behavior is controlled by the `dotnet_style_operator_placement_when_wrapping` option in **.editorconfig**. The default value is `beginning_of_line`, but it can be set to `end_of_line` if desired.

**Comment by Timovzl** [comment]
For future readers, I believe this deserves the nuance of a counterargument. A reason why one might _not_ prefer what the original post suggests is the following:

To reduce cognitive load and improve ease of understandability, it is desirable when **equivalent thing are presented in a way that makes them _look_ equivalent**. For example:

```cs
// Unbalanced
var selection = collection.Select(element => element.Value) // Why am I treated differently from my peers?
	.Where(value => value is not null)
	.ToList();

// Balanced
var selection = collection
	.Select(element => element.Value)
	.Where(value => value is not null)
	.ToList();
```

Following this reasoning, the suggested formatting for the ternary operator is indeed desirable:

```cs
// Balanced
this.Key = Enum.TryParse<Key>(key, ignoreCase: true, out var keyboardKey) // Condition
	? keyboardKey // Consequence
	: null; // Consequence
```

As we can see, the consequences are aligned. The condition is not, but it is different concept, so this makes sense.

However, the suggested formatting for binary operations is then _undesirable_:

```cs
// Unbalanced
var foo = barValue // Why do I have different indentation than my peers?
	&& checkBazValue()
	&& qux.IsTrue();

// Unbalanced
this.Character.Location += this.DirectionMode.GetDeltaPoint() // Why do I have different indentation than my peers?
	* (float)interval.TotalMilliseconds 
	* this.Speed;

// Unbalaced, even with an extra line:
var foo =
	barValue // Why do I have different indentation than my peers?
	&& checkBazValue()
	&& qux.IsTrue();
```

With the `end_of_line` style, this can be cured:

```cs
// Balanced
var foo =
	barValue &&
	checkBazValue() &&
	qux.IsTrue();

// Balanced
this.Character.Location +=
	this.DirectionMode.GetDeltaPoint() *
	(float)interval.TotalMilliseconds *
	this.Speed;
```

Note that although proper alignment can technically also be achieved with the `beginning_of_line` style, that requires custom indentation. Doing so is rather cumbersome, and it tends to get undone by Format Document.

```cs
// Cumbersome, fragile
var foo =
	   barValue
	&& checkBazValue()
	&& qux.IsTrue();
```

**In the end, the trade-off is in what vertical scanning should provide:**
- `beginning_of_line`: **Quick insight into the relationship between the elements (i.e. the operators).**
- `end_of_line`: **Quick insight into the structure as a whole (with equivalent things _looking_ equivalent).**
